x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,1301 @@
+- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
+  - Used for attaching shared exceptions list to rules
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
+  - Deprecated and unused by the UI (might still be used by API users)
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
+  - Deprecated and unused by the UI (might still be used by API users)
+- **Import Rules** - `POST /rules/_import`: 
+  - See section [Importing rules](#importing-rules){#toc-importing-rules}
+- **Bulk Actions** - `POST /rules/_bulk_action`: 
+    - This endpoint includes many types of actions, but only the **bulk edit**  and the **duplicate** actions should perform a migration on write.
+    - This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`, since we never write to ES when this parameter is set to `true`.
+    - For the **bulk edit** action: 
+      - we can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.
+      - See implementation details below.
+    - For the **duplicate** rule action:
+      - since we will be creating a new rule on ES, we should create it with the new schema. 
+      - Per definition: all duplicated rules will be `custom` rules. That means that for all rules -including prebuilt rules-, when duplicating, we need to set the new rule's `prebuilt`, `immutable` and `customized` value to `false`.
+      - however, this action will not perform a migration-on-write of the original rule being duplicated for two reasons:
+        - would violate the principle of least surprise for the endpoint
+        - the current implementation of the endpoint does not modify the original rule. This gives us no window of opportunity to migrate the rule and save to ES without adding a considerable performance overhead.
+      - See implementation details below.
+    - All other type of actions should **not** perform migration-on-write:
+        - Enable
+        - Disable
+        - Delete
+        - Export: see Importing and Exporting Rules section below
+
+**NOTE:** Endpoints that depend on the Rules Client's methods need to modify the rules' params (prebuilt and customized) before the Rule Client methods call the`validateRuleTypeParams`, because that util validates the passed parameters using the validators defined in `x-pack/plugins/security_solution/server/lib/detection_engine/rule_types`. Since those are an intersection between `BaseRulesParams` and the specific rule type's fields, `prebuilt` and `customized` are non-optional, which would cause the util to reject the validation of the params for non-migrated rules.
```

Comment:
`validateRuleTypeParams` is also called when rules run, which is another reason that the internal rule schema must not specify `prebuilt` and `customized` as required if we're doing migration on write. This will cause rules to fail at runtime.